kgforge/core/commons/sparql_query_builder.py

```diff
@@ -345,20 +367,120 @@ def _replace_in_sparql(
+                re.compile(r" LIMIT \d+", flags=re.IGNORECASE),
             )
         if offset:
             query = SPARQLQueryBuilder._replace_in_sparql(
-                query, "OFFSET", offset, default_offset,
-                re.compile(r" OFFSET \d+", flags=re.IGNORECASE)
+                query,
+                "OFFSET",
+                offset,
+                default_offset,
+                re.compile(r" OFFSET \d+", flags=re.IGNORECASE),
             )
 
         return query
 
+    @staticmethod
+    def create_select_query(
+        vars_: List[str],
+        statements: Union[str, List[str]],
+        distinct: bool,
+        search_in_graph: bool,
+        union: bool = False,
+    ):
+        if not union:
+            all_statements = (
+                ". ".join(statements) if isinstance(statements, list) else statements
+            )
+
+        elif not isinstance(statements, list):
+            raise AttributeError(
+                f"Expected statements argument to be of type  List[str] when union is True. Each item will go in a UNION clause: {statements}"
+            )
+        else:
+            all_statements = "}} UNION {{".join(statements)
+            all_statements = "{{" + all_statements + "}}"
+
+        where_clauses = (
+            f"{{ Graph ?g {{{all_statements}}}}}"
+            if search_in_graph
+            else f"{{{all_statements}}}"
+        )
+
+        join_vars_ = " ".join(vars_)
+        select_vars = f"DISTINCT {join_vars_}" if distinct else f"{join_vars_}"
+        return f"SELECT {select_vars} WHERE {where_clauses}"
+
 
 def _box_value_as_full_iri(value):
     return f"<{value}>" if is_valid_url(value) else value
+
+
+def build_shacl_query(
+    statements: List[str] = None,
+    defining_property_uri: str = None,
+    deprecated_property_uri: str = None,
+    deprecated: bool = False,
+    deprecated_optional: bool = False,
+    search_in_graph: bool = False,
+    context: Context = None,
+) -> str:
+    deprecated_statement = (
+        f"?resource_id <{deprecated_property_uri}> '{str(deprecated).lower()}'^^xsd:boolean"
+        if deprecated_property_uri
+        else ""
+    )
+    defining_statement = (
+        f"?resource_id <{defining_property_uri}> ?shape"
+        if defining_property_uri
+        else ""
+    )
+    if deprecated_optional and deprecated_property_uri:
+        deprecated_statement = " OPTIONAL {" + deprecated_statement + "}"
+    shape_target_statement = "?shape sh:targetClass ?type"
+    extra_statements = [defining_statement, deprecated_statement]
+    if statements:
+        extra_statements.extend(statements)
+    shape_target_statement_str = ".".join([shape_target_statement] + extra_statements)
+    shape_node_statements_str = (
+        "SELECT (?shape as ?type) ?shape ?resource_id WHERE { ?shape a sh:NodeShape ."
+        + "?shape a rdfs:Class . {}".format(".".join(extra_statements) + "}")
+    )
+    shape_node_not_targeting_statements_str = (
```

Comment:
this statement doesn't seem to return ?type, and then it fails in the tests for 
graph: /tests/data/shacl-model/commons/shapes-3.json
schema: "http://shapes.ex/building"
shape: GeoShape

Not sure if the issue is the shape or the query though

Reply:
This should be fix now